pretraining/utils/dataset_utils.py

```diff
@@ -821,6 +832,19 @@ def load_state_dict(self, state_dicts, sharded_input=False):
         return super().load_state_dict(state_dicts, sharded_input)
 
 
+class _Stateful_Dataset_With_Datasets(Protocol):
+    def __init__(self, rank: int, worldsize: int, datasets: List[Any], **kwargs):
+        pass
+
+
+class _With_Docset(Protocol):
+    docset: List[Any]
+
+
+class _Stateful_Dataset_With_Docset(_Stateful_Dataset, _With_Docset):
+    pass
+
```

Comment:
I don't know the purpose behind the code well enough to tell, but the fact the there is a dataset type argument in these dataset classes seems to suggest that different dataset types may be passed, and there is no base class today that satisfies all requirements. 

The classes I added solve that problem, but it is also possible to amend one the stateful dataset base class and add datasets and docset there, which would eliminate the need for these extra classes.

Reply:
in the doc for `Sampling_Dataset` (first use-case I see for this class) the docs say:
`dataset : Shard_Line_Dataset | Shard_Doc_Dataset | Streaming_Doc_Dataset`

This is all code that was adapted from internal code, and some of those classes referenced in the code comment might not still be present here, or may be deprecated. If they're all here, we could use a Union type without changing other code. In practice I think we'll end up only finding one `dataset` class being used, so we could just use the singular class that's actually passed within this repo.

If there's really only one, the code may get changed to not even have this as a parameter, but as much as possible I'd rather avoid including refactoring with the mypy commit.

Reply:
Right now, the code will accept anything that is a subclass of `_Stateful_Dataset`, with my change the code will accept anything that is a subclass of `_Stateful_Dataset` as long as the constructor includes a `datasets` argument. 

If we use a `Union` it will be a stricter restriction since it will only allow the list of classes in the `Union`, but I'm happy to do it if it's your preferred approach. It will still be possible to add to the `Union` or introduce a different parent class later.